detector.py

```diff
@@ -1,34 +1,102 @@
+    def detect(self, img, length, decimal, thr=0.4):
+        img =  self.detect_plates(img)
+        img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
+        return None if img is None else self.detect_numbers(img, length, decimal, thr)
 
     def detect_plates(self, img):
-        imgs = []
-        results = self.plates_model.predict(source=[img])
-        print(f"Detector found {len(results)} number plates")
-        for result in results:
-            for box in result.boxes:
-                x1, y1, x2, y2 = np.round(box.xyxy[0].numpy()).astype(int)
-                imgs.append(img[y1:y2, x1:x2])
-        return imgs
-
-    def detect_numbers(self, img, length, decimal):
-        _, im_buf_arr = cv2.imencode(".jpg", img)
-        byte_im = im_buf_arr.tobytes()
-
-        return detect_text(byte_im, length)
+        self.debug_img = img
+        result = self.plates_model.predict(source=[img], verbose=False)[0]
+        Logger.debug(f"Detected {len(result)} plate(s).")
+        result.boxes.data = sorted(result.boxes.data, key=lambda x: x[4], reverse=True)
+
+        for box in result.boxes:
+            x1, y1, x2, y2 = np.round(box.xyxy[0].numpy()).astype(int)
+            self.debug_img = [x1, y1, result.plot()]
+            return img[y1:y2, x1:x2]
+        return None
+
+    def detect_numbers(self, img, length, decimal, thr):
+        img = self.automatic_brightness_and_contrast(img)
+        result = self.number_model.predict(source=[img], verbose=False)[0]
+        (f"Detected {len(result)} number(s).")
+
+        # Remove below threshold
+        detections = []
+        r = 0
+        for row in result.boxes.data:
+            if row[4] >= thr:
+                detections.append(row)
+            else: r += 1
+        if r > 0: Logger.debug(f"Removed {r} image(s) below threshold ({thr}).")
+
+        # Debug plot
+        self.debug_img[2][self.debug_img[1]:self.debug_img[1]+result.plot().shape[0], self.debug_img[0]:self.debug_img[0]+result.plot().shape[1]] = result.plot()
+        self.debug_img = self.debug_img[2]
+        cv2.imwrite("library/debug.png", self.debug_img)
+        Logger.debug("Debug image saved.")
+
+        # Not enough numbers found
+        if len(detections) < length:
+            Logger.debug("Not enough numbers detected for valid number.")
+            return None
+
+        # Sort by size and use first "length"s numbers
+        r = len(detections) - length
+        if r > 0: Logger.debug(f"Removed {r} less relevant number(s) from detection.")
+        detections = sorted(detections, key=lambda x: x[2]*x[3], reverse=False)[:length]
+
+        # Sort by X coordinate
+        detections = sorted(detections, key=lambda x: x[0], reverse=False)
+
+        # Create float number from list
+        nums = [str(int(row[-1])) for row in detections]
+        return int(''.join(nums)) / (10** decimal)
+
+    def automatic_brightness_and_contrast(self, image, clip_hist_percent=1):
```

Comment:
wouldnt be better to put this into a separate file like cm_utils or cm_cv_functions or something like this ?